### ISSUE TITLE
Compiled regular expressions cache is exhausted by language detection

### ISSUE THREAD

**Issue by ArturGaspar** [scrapinghub/dateparser#144]
The `re` module's cache of compiled regular expressions is exhausted by the language detection when it fails to detect a language, causing the same regular expressions to be compiled again each time an invalid date is passed.
```
>>> import timeit
>>> from dateparser import DateDataParser
>>> ddp = DateDataParser(allow_redetect_language=True)
>>> timeit.timeit(lambda: ddp.get_date_data("not really a date"), number=100)
13.68727993965149
```

Increasing `re._MAXCACHE` prevents the cache from being cleared and all the patterns from being compiled again on each attempt.
```
>>> import timeit
>>> from dateparser import DateDataParser
>>> ddp = DateDataParser(allow_redetect_language=True)
>>> import re
>>> re._MAXCACHE = float('inf')
>>> timeit.timeit(lambda: ddp.get_date_data("not really a date"), number=100)
3.077785015106201
```

Profiling shows that most time is spent in `re.sub()` called by `Language._simplify()` and `re.compile()` called by `Dictionary._construct_split_regex()`.

**Comment by waqasshabbir** [comment]
Thanks @ArturGaspar! will look into it.

**Comment by asadurski** [comment]
Thank you, @ArturGaspar. I'd like to ask you a favor. Could you check the performance with `regex` module, which has some improvements for caching? It's introduced in not yet merged PR #105.

**Comment by ArturGaspar** [comment]
@asadurski It's much faster, faster even than using `re` with that workaround for the cache size.

**Comment by waqasshabbir** [comment]
Fixed in PR #105.
